### ISSUE TITLE
xterm mouse support: only 223 chars wide?

### ISSUE THREAD

**Issue by jafcobend** [cosmos72/twin#72]
Using `twin --hw=tty` in xterm over an SSH connection to a remote machine (twin on the remote side) it seems that the mouse stops working at about 223 chars from the left. I don't have this problem with hw_X11 on the local system. Is this an issue with the xterm/gpm mouse protocol? Or something broken in twin?

**Comment by cosmos72** [comment]
It makes sense.
Basic xterm mouse protocol uses a single byte >= 32 to represent each coordinate, so it's limited to approx 255-32 = 223 in both maximum x and maximum y.
Honestly, I don't know if xterm also supports some more advanced mouse protocol with higher limits.

**Comment by jafcobend** [comment]
OK. That makes sense then. It doesn't appear `gpm` has the same issue. Running on a native Linux console at full HD res I can use the mouse out 230+ chars. I suppose that's why it surprised me. I'm not sure how many others would do something like this. ;-) Thanks for the info.

**Comment by cosmos72** [comment]
Update:
https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h2-Mouse-Tracking
states that xterm also supports an extended mouse protocol with maximum height and width = 2095.
I will have a look

**Comment by jafcobend** [comment]
**SWEET!** I'll have to read up on that. I don't suppose it has scroll-wheel support? TWIN has become my replacement for `tmux` and `screen`. Being able to expand my session window to cover one or both my HD monitors is sometimes very useful. 8-)

**Comment by cosmos72** [comment]
Probably yes - scroll wheel is usually treated as two additional buttons.

**Comment by cosmos72** [comment]
I have implemented both in commit e91ef2283ff982b2823949057e73dfdfad0d4b0e:
> server/hw/hw_tty: improve mouse support on xterm-like terminals
> * add support for mouse scroll wheel
> * use enhanced xterm mouse protocol ESC [?1002h ESC [?1006h to support mouse positions > 223 (both in x and y)

**Comment by jafcobend** [comment]
Sweet! Thanks! I'm pulling them down now and will start the baking process. I'm working on some mail logs on a remote server so the extra width will be **REALLY** nice!

**Comment by jafcobend** [comment]
**BINGO** It works! Now I need to try it some other terminal environment to see if anything got broken....

**Comment by jafcobend** [comment]
Scroll wheel works through xterm too!

**Comment by cosmos72** [comment]
Just for information, a setup that always works perfectly is:
Run twin locally;
Open a terminal inside twin and ssh to a remote host;
In the remote host, run `twin --hw=tty`

**Comment by jafcobend** [comment]
> Just for information, a setup that always works perfectly is:
> Run twin locally (typically with --hw=X or --hw=xft)

Not really. It has catastrophic failures. I'll have to play with it a bit more. But at minimum with --hw=X11 locally when the remote session closes the menu bar goes away and the keyboard becomes inoperable. Then the local TWIN has to be closed to work again.

As I said I'll have to play with it more to quantify exactly what's going wrong and what the specific symptoms are. But that was my initial impression. Right now xterm to a remote TWIN works well... unless you're reestablishing a disconnected session. I'll have to play with that a bit more too...

**Comment by cosmos72** [comment]
This is *really* strange, as it's a configuration that I use often.
If you can investigate it further, I'd really appreciate that (please open a separate issue when you report what you found)

**Comment by jafcobend** [comment]
I can't get it to break again. Maybe getting all of my installs up to the latest version helped. Or maybe it was a slightly different manifestation of a bug I've just started seeing in the combination of X server, power management and xscreensaver in this new OS _upgrade_ I just started using. But if I stumble into it again I'll make sure to poke at it more, quantify it and open an issue. :-D
